File: backend/utils/safety.js
const path = require("node:path");
const os = require("node:os");
const logger = require("../utils/logger");

class SafetyService {
  constructor() {
    this.initializeWhitelists();
  }

  initializeWhitelists() {
    // 允许的应用程序白名单
    this.allowedApps = [
      "spotify",
      "music",
      "itunes",
      "chrome",
      "safari",
      "firefox",
      "visual studio code",
      "vscode",
      "terminal",
      "finder",
      "calculator",
      "calendar",
      "mail",
      "notes",
      "pages",
      "numbers",
      "keynote",
      "preview",
      "photos",
      "system preferences",
      "activity monitor",
      "textedit",
      "reminder",
      "maps",
      "weather",
      "contacts",
    ];

    // 允许的目录白名单
    this.allowedDirectories = [
      path.join(os.homedir(), "Documents"),
      path.join(os.homedir(), "Desktop"),
      path.join(os.homedir(), "Downloads"),
      path.join(os.homedir(), "Music"),
      path.join(os.homedir(), "Pictures"),
      path.join(os.homedir(), "Movies"),
      path.join(os.homedir(), "Documents/VoiceAssistant"),
      "/tmp",
      "/var/tmp",
    ];

    // 危险命令黑名单
    this.dangerousCommands = [
      "rm -rf /",
      "sudo rm",
      "format",
      "fdisk",
      "mkfs",
      "shutdown",
      "reboot",
      "halt",
      "poweroff",
      "init 0",
      "killall",
      "pkill",
      "kill -9",
      "chmod 777",
      "chown",
      "passwd",
      "su",
      "sudo su",
      "crontab",
      "at",
      "batch",
      "nohup",
      "dd if=",
      "cat /dev/",
      "echo > /dev/",
      "mount",
      "umount",
      "fsck",
      "iptables",
      "ufw",
      "firewall",
      "route",
      "arp",
      "netstat",
      "sshd",
    ];

    // 危险文件扩展名
    this.dangerousExtensions = [
      ".exe",
      ".bat",
      ".cmd",
      ".com",
      ".pif",
      ".scr",
      ".vbs",
      ".js",
      ".jar",
      ".app",
      ".dmg",
      ".pkg",
      ".mpkg",
      ".deb",
      ".rpm",
      ".run",
      ".bin",
    ];

    // 敏感系统路径
    this.systemPaths = [
      "/bin",
      "/sbin",
      "/usr/bin",
      "/usr/sbin",
      "/etc",
      "/var",
      "/sys",
      "/proc",
      "/dev",
      "/boot",
      "/lib",
      "/lib64",
      "/usr/lib",
      "/usr/lib64",
    ];
  }

  async validateToolCall(toolCall, context = {}) {
    try {
      const { name, arguments: args } = toolCall;

      logger.info(`安全校验工具调用: ${name}`, { args, context });

      if (context && context.allowLocalControl === false) {
        const localControlTools = ["write_file", "open_app", "file_control", "write_run_code", "send_message"];
        if (localControlTools.includes(name)) {
          return {
            allowed: false,
            riskLevel: "high",
            reason: "本地应用操控已被关闭",
            requiresConfirmation: false,
          };
        }
      }

      // 基础参数校验
      const basicValidation = this.validateBasicParameters(name, args);
      if (!basicValidation.valid) {
        return {
          allowed: false,
          riskLevel: "high",
          reason: basicValidation.reason,
          requiresConfirmation: true,
        };
      }

      // 根据工具类型进行特定校验
      let specificValidation;
      switch (name) {
        case "play_music":
          specificValidation = this.validatePlayMusic(args);
          break;
        case "stop_music":
          specificValidation = this.validateStopMusic(args);
          break;
        case "write_article":
          specificValidation = this.validateWriteArticle(args);
          break;
        case "write_file":
          specificValidation = this.validateWriteFile(args);
          break;
        case "open_app":
          specificValidation = this.validateOpenApp(args);
          break;
        case "send_message":
          specificValidation = this.validateSendMessage(args);
          break;
        case "write_run_code":
          specificValidation = this.validateWriteRunCode(args);
          break;
        case "file_control":
          specificValidation = this.validateFileControl(args);
          break;
        default:
          specificValidation = { valid: false, reason: `未知的工具类型: ${name}` };
      }

      if (!specificValidation.valid) {
        return {
          allowed: false,
          riskLevel: "high",
          reason: specificValidation.reason,
          requiresConfirmation: true,
        };
      }

      // 评估风险等级
      const riskAssessment = this.assessRisk(name, args, context);

      return {
        allowed: true,
        riskLevel: riskAssessment.level,
        reason: riskAssessment.reason,
        requiresConfirmation: riskAssessment.requiresConfirmation,
        suggestions: riskAssessment.suggestions,
      };
    } catch (error) {
      logger.error("安全校验失败:", error);
      return {
        allowed: false,
        riskLevel: "high",
        reason: `安全校验过程出错: ${error.message}`,
        requiresConfirmation: true,
      };
    }
  }

  validateBasicParameters(name, args) {
    if (!name || typeof name !== "string") {
      return { valid: false, reason: "工具名称无效" };
    }

    if (!args || typeof args !== "object") {
      return { valid: false, reason: "工具参数无效" };
    }

    // 检查参数中是否包含危险内容 - 使用更精确的匹配
    const argsString = JSON.stringify(args).toLowerCase();
    for (const dangerousCmd of this.dangerousCommands) {
      // 优先用更稳健的匹配：对包含特殊字符的命令使用子串匹配，避免 \b 边界导致漏检
      const loweredCmd = String(dangerousCmd).toLowerCase();
      const needsSimpleMatch = /[^a-z0-9_\s]/i.test(dangerousCmd);

      if (needsSimpleMatch) {
        if (argsString.includes(loweredCmd)) {
          return { valid: false, reason: `参数包含危险命令: ${dangerousCmd}` };
        }
        continue;
      }

      // 对纯“单词/空格”命令使用边界匹配，降低误报
      const escapedCmd = dangerousCmd.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      const regex = new RegExp(`\\b${escapedCmd}\\b`, "i");
      if (regex.test(argsString)) {
        return { valid: false, reason: `参数包含危险命令: ${dangerousCmd}` };
      }
    }

    return { valid: true };
  }

  validatePlayMusic(args) {
    const { source, query } = args;

    // source 可选：缺省时由后端自动选择可用播放器/来源
    const allowedSources = ["spotify", "apple", "local"];
    if (source && !allowedSources.includes(source)) {
      return { valid: false, reason: `不支持的音乐来源: ${source}` };
    }

    if (query && typeof query !== "string") {
      return { valid: false, reason: "查询参数必须是字符串" };
    }

    if (query && query.length > 200) {
      return { valid: false, reason: "查询参数过长" };
    }

    return { valid: true };
  }

  validateStopMusic(_args) {
    // stop_music 不需要参数，总是安全的
    return { valid: true };
  }

  validateWriteArticle(args) {
    const { topic, style, length } = args;

    if (!topic) {
      return { valid: false, reason: "缺少文章主题参数" };
    }

    if (typeof topic !== "string" || topic.length > 500) {
      return { valid: false, reason: "文章主题无效或过长" };
    }

    if (style && !["formal", "casual", "professional", "creative"].includes(style)) {
      return { valid: false, reason: `不支持的写作风格: ${style}` };
    }

    if (length && !["short", "medium", "long"].includes(length)) {
      return { valid: false, reason: `不支持的文章长度: ${length}` };
    }

    return { valid: true };
  }

  validateWriteFile(args) {
    const { path: filePath, content, mode } = args;

    if (!filePath) {
      return { valid: false, reason: "缺少文件路径参数" };
    }

    if (typeof filePath !== "string") {
      return { valid: false, reason: "文件路径必须是字符串" };
    }

    // 检查路径遍历攻击
    if (filePath.includes("..") || filePath.includes("~")) {
      return { valid: false, reason: "文件路径包含不安全字符" };
    }

    // 检查是否在允许的目录内
    const absolutePath = path.isAbsolute(filePath) ? filePath : path.join(os.homedir(), filePath);
    const isAllowed = this.allowedDirectories.some((allowedDir) =>
      absolutePath.startsWith(allowedDir)
    );

    if (!isAllowed) {
      return { valid: false, reason: "文件路径不在允许的目录范围内" };
    }

    // 检查文件扩展名
    const ext = path.extname(filePath).toLowerCase();
    if (this.dangerousExtensions.includes(ext)) {
      return { valid: false, reason: `不允许的文件类型: ${ext}` };
    }

    if (content && typeof content !== "string") {
      return { valid: false, reason: "文件内容必须是字符串" };
    }

    if (content && content.length > 1000000) {
      // 1MB限制
      return { valid: false, reason: "文件内容过大" };
    }

    if (mode && !["create", "append", "overwrite"].includes(mode)) {
      return { valid: false, reason: `不支持的写入模式: ${mode}` };
    }

    return { valid: true };
  }

  validateOpenApp(args) {
    const { name } = args;

    if (!name) {
      return { valid: false, reason: "缺少应用程序名称参数" };
    }

    if (typeof name !== "string") {
      return { valid: false, reason: "应用程序名称必须是字符串" };
    }

    const trimmedName = name.trim();
    if (!trimmedName) {
      return { valid: false, reason: "应用程序名称不能为空" };
    }

    if (trimmedName.length > 200) {
      return { valid: false, reason: "应用程序名称过长" };
    }

    // 禁止用户直接传路径，统一走“已安装应用检索”逻辑
    if (trimmedName.includes("/") || trimmedName.includes("\\")) {
      return { valid: false, reason: "应用程序名称不应包含路径分隔符" };
    }

    // 基础注入防护：禁止控制字符与明显的 shell 元字符
    if (/\r|\n|\0/.test(trimmedName) || /[;&|<>]/.test(trimmedName)) {
      return { valid: false, reason: "应用程序名称包含不安全字符" };
    }

    return { valid: true };
  }

  validateSendMessage(args) {
    const { target, content, channel } = args;

    if (!target) {
      return { valid: false, reason: "缺少消息目标参数" };
    }

    if (typeof target !== "string" || target.trim().length > 200) {
      return { valid: false, reason: "消息目标无效或过长" };
    }

    if (!content) {
      return { valid: false, reason: "缺少消息内容参数" };
    }

    if (typeof content !== "string" || content.length > 5000) {
      return { valid: false, reason: "消息内容无效或过长" };
    }

    if (channel && !["auto", "sms", "email", "im"].includes(channel)) {
      return { valid: false, reason: `不支持的发送渠道: ${channel}` };
    }

    return { valid: true };
  }

  validateWriteRunCode(args) {
    const { language, code, run } = args;

    if (!language || !["javascript", "python", "bash"].includes(language)) {
      return { valid: false, reason: "代码语言不支持或缺失" };
    }

    if (!code || typeof code !== "string" || code.length > 50000) {
      return { valid: false, reason: "代码内容无效或过长" };
    }

    if (typeof run !== "undefined" && typeof run !== "boolean") {
      return { valid: false, reason: "run 参数必须是布尔值" };
    }

    return { valid: true };
  }

  isPathAllowed(filePath) {
    if (!filePath || typeof filePath !== "string") {
      return { allowed: false, reason: "文件路径无效" };
    }

    if (filePath.includes("..") || filePath.includes("~")) {
      return { allowed: false, reason: "文件路径包含不安全字符" };
    }

    const absolutePath = path.isAbsolute(filePath) ? filePath : path.join(os.homedir(), filePath);
    const isAllowed = this.allowedDirectories.some((allowedDir) => absolutePath.startsWith(allowedDir));

    if (!isAllowed) {
      return { allowed: false, reason: "文件路径不在允许的目录范围内" };
    }

    const ext = path.extname(filePath).toLowerCase();
    if (ext && this.dangerousExtensions.includes(ext)) {
      return { allowed: false, reason: `不允许的文件类型: ${ext}` };
    }

    return { allowed: true };
  }

  validateFileControl(args) {
    const { operation, path: filePath, destination } = args;

    if (!operation || !["list", "read", "move", "copy", "delete", "mkdir"].includes(operation)) {
      return { valid: false, reason: "文件操作类型不支持或缺失" };
    }

    const pathCheck = this.isPathAllowed(filePath);
    if (!pathCheck.allowed) {
      return { valid: false, reason: pathCheck.reason };
    }

    if (destination) {
      const dstCheck = this.isPathAllowed(destination);
      if (!dstCheck.allowed) {
        return { valid: false, reason: `目标路径不安全: ${dstCheck.reason}` };
      }
    }

    return { valid: true };
  }

  assessRisk(toolName, args, _context) {
    // 基础风险评估
    const riskLevels = {
      play_music: { level: "low", requiresConfirmation: false },
      stop_music: { level: "low", requiresConfirmation: false },
      open_app: { level: "low", requiresConfirmation: false },
      write_article: { level: "medium", requiresConfirmation: false },
      write_file: { level: "medium", requiresConfirmation: true },
      send_message: { level: "medium", requiresConfirmation: true },
      write_run_code: { level: "high", requiresConfirmation: true },
      file_control: { level: "high", requiresConfirmation: true },
    };

    const baseRisk = riskLevels[toolName] || { level: "medium", requiresConfirmation: true };

    // 根据参数调整风险等级
    const adjustedRisk = { ...baseRisk };

    if (toolName === "write_file") {
      const { path: filePath, mode } = args;

      // 覆盖模式风险更高
      if (mode === "overwrite") {
        adjustedRisk.level = "high";
        adjustedRisk.requiresConfirmation = true;
      }

      // 系统路径风险更高
      if (typeof filePath === "string" && filePath.trim()) {
        const absolutePath = path.isAbsolute(filePath) ? filePath : path.join(os.homedir(), filePath);
        if (this.systemPaths.some((sysPath) => absolutePath.startsWith(sysPath))) {
          adjustedRisk.level = "high";
          adjustedRisk.requiresConfirmation = true;
        }
      }
    }

    if (toolName === "open_app") {
      const { name } = args;

      // 系统工具风险更高
      const systemApps = ["terminal", "activity monitor", "system preferences"];
      if (systemApps.includes(name.toLowerCase())) {
        adjustedRisk.level = "medium";
        adjustedRisk.requiresConfirmation = true;
      }
    }

    // 生成风险说明和建议
    const riskDescriptions = {
      low: "低风险操作，可以安全执行",
      medium: "中等风险操作，建议确认后执行",
      high: "高风险操作，必须用户确认才能执行",
    };

    return {
      level: adjustedRisk.level,
      requiresConfirmation: adjustedRisk.requiresConfirmation,
      reason: riskDescriptions[adjustedRisk.level],
      suggestions: this.generateSuggestions(toolName, args, adjustedRisk.level),
    };
  }

  generateSuggestions(toolName, args, riskLevel) {
    const suggestions = [];

    if (toolName === "write_file") {
      const { path: filePath, mode } = args;
      suggestions.push(
        `将${mode === "append" ? "追加到" : mode === "overwrite" ? "覆盖" : "创建"}文件: ${filePath}`
      );

      if (riskLevel === "high") {
        suggestions.push("此操作可能覆盖现有文件，请确认是否继续");
      }
    }

    if (toolName === "open_app") {
      const { name } = args;
      suggestions.push(`将打开应用程序: ${name}`);
    }

    if (toolName === "play_music") {
      const { source, query } = args;
      const sourceText = source ? `${source} ` : "";
      const autoHint = source ? "" : "（自动选择播放器）";
      suggestions.push(`将播放${sourceText}音乐${query ? `: ${query}` : ""}${autoHint}`);
    }

    if (toolName === "send_message") {
      const { target, channel } = args;
      const channelText = channel && channel !== "auto" ? `（渠道: ${channel}）` : "";
      suggestions.push(`将发送消息给: ${target}${channelText}`);
    }

    if (toolName === "write_run_code") {
      const { language, run } = args;
      suggestions.push(`将编写${language}代码${run ? "并运行" : ""}`);
      if (riskLevel === "high") {
        suggestions.push("此操作可能执行本地代码，请确认是否继续");
      }
    }

    if (toolName === "file_control") {
      const { operation, path: filePath, destination } = args;
      const dstText = destination ? ` → ${destination}` : "";
      suggestions.push(`将执行文件操作: ${operation} ${filePath}${dstText}`);
      if (riskLevel === "high") {
        suggestions.push("此操作可能影响本地文件，请确认是否继续");
      }
    }

    return suggestions;
  }

  generateConfirmationRequest(toolCall, riskAssessment) {
    return {
      id: `confirm_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      toolCall: {
        id: toolCall.id,
        name: toolCall.name,
        arguments: toolCall.arguments,
      },
      riskLevel: riskAssessment.level,
      summary: this.generateSummary(toolCall),
      reason: riskAssessment.reason,
      suggestions: riskAssessment.suggestions,
      timestamp: new Date().toISOString(),
      expiresAt: new Date(Date.now() + 5 * 60 * 1000).toISOString(), // 5分钟后过期
    };
  }

  generateSummary(toolCall) {
    const { name, arguments: args } = toolCall;

    switch (name) {
      case "play_music": {
        const sourceText = args.source ? `${args.source} ` : "";
        const autoHint = args.source ? "" : "（自动选择播放器）";
        return `播放${sourceText}音乐${args.query ? `: ${args.query}` : ""}${autoHint}`;
      }
      case "stop_music":
        return "停止音乐播放";
      case "write_article":
        return `写文章: ${args.topic}`;
      case "write_file":
        return `${args.mode === "append" ? "追加到" : args.mode === "overwrite" ? "覆盖" : "创建"}文件: ${args.path}`;
      case "open_app":
        return `打开应用程序: ${args.name}`;
      case "send_message":
        return `发送消息给: ${args.target}`;
      case "write_run_code":
        return `编写${args.language}代码${args.run ? "并运行" : ""}`;
      case "file_control": {
        const dstText = args.destination ? ` → ${args.destination}` : "";
        return `文件操作: ${args.operation} ${args.path}${dstText}`;
      }
      default:
        return `执行操作: ${name}`;
    }
  }

  // 更新白名单的方法
  updateAllowedApps(apps) {
    this.allowedApps = [...new Set([...this.allowedApps, ...apps])];
    logger.info("更新允许应用白名单", { apps });
  }

  updateAllowedDirectories(dirs) {
    this.allowedDirectories = [...new Set([...this.allowedDirectories, ...dirs])];
    logger.info("更新允许目录白名单", { dirs });
  }

  // 获取当前配置
  getSafetyConfig() {
    return {
      allowedApps: this.allowedApps,
      allowedDirectories: this.allowedDirectories,
      dangerousCommands: this.dangerousCommands,
      dangerousExtensions: this.dangerousExtensions,
      systemPaths: this.systemPaths,
    };
  }
}

module.exports = SafetyService;

File: backend_py/safety.py
from __future__ import annotations

import os
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List, Optional


@dataclass
class RiskAssessment:
    allowed: bool
    risk_level: str
    reason: str
    requires_confirmation: bool


class SafetyService:
    """安全校验（与 `backend/utils/safety.js` 行为对齐的最小移植版）。"""

    def __init__(self) -> None:
        home = str(Path.home())
        self.allowed_apps = {
            "spotify",
            "music",
            "itunes",
            "chrome",
            "safari",
            "firefox",
            "visual studio code",
            "vscode",
            "terminal",
            "finder",
            "calculator",
            "calendar",
            "mail",
            "notes",
            "pages",
            "numbers",
            "keynote",
            "preview",
            "photos",
            "system preferences",
            "activity monitor",
            "textedit",
            "reminder",
            "maps",
            "weather",
            "contacts",
        }

        self.allowed_directories = [
            os.path.join(home, "Documents"),
            os.path.join(home, "Desktop"),
            os.path.join(home, "Downloads"),
            os.path.join(home, "Music"),
            os.path.join(home, "Pictures"),
            os.path.join(home, "Movies"),
            os.path.join(home, "Documents", "VoiceAssistant"),
            "/tmp",
            "/var/tmp",
        ]

        self.dangerous_extensions = {
            ".exe",
            ".bat",
            ".cmd",
            ".com",
            ".pif",
            ".scr",
            ".vbs",
            ".js",
            ".jar",
            ".app",
            ".dmg",
            ".pkg",
            ".mpkg",
            ".deb",
            ".rpm",
            ".run",
            ".bin",
        }

    def _is_path_allowed(self, file_path: str) -> RiskAssessment:
        if not file_path or not isinstance(file_path, str):
            return RiskAssessment(False, "high", "文件路径无效", True)

        if ".." in file_path or "~" in file_path:
            return RiskAssessment(False, "high", "文件路径包含不安全字符", True)

        p = Path(file_path)
        abs_path = p if p.is_absolute() else (Path.home() / p)
        abs_path_str = str(abs_path)

        is_allowed = any(abs_path_str.startswith(d) for d in self.allowed_directories)
        if not is_allowed:
            return RiskAssessment(False, "high", "文件路径不在允许的目录范围内", True)

        ext = p.suffix.lower()
        if ext and ext in self.dangerous_extensions:
            return RiskAssessment(False, "high", f"不允许的文件类型: {ext}", True)

        return RiskAssessment(True, "medium", "ok", True)

    def validate_tool_call(self, tool_call: Dict[str, Any], *, allow_local_control: bool = True) -> RiskAssessment:
        name = tool_call.get("name")
        args = tool_call.get("arguments") or {}

        if not isinstance(name, str) or not name:
            return RiskAssessment(False, "high", "工具名无效", False)

        local_control_tools = {
            "write_file",
            "open_app",
            "file_control",
            "write_run_code",
            "send_message",
            "run_tests",
            "execute_workflow",
            "music_ui",
            "media_control",
        }
        if allow_local_control is False and name in local_control_tools:
            return RiskAssessment(False, "high", "本地应用操控已被关闭", False)

        # 按工具类型做参数与风险校验
        if name == "open_app":
            app_name = args.get("name")
            if not app_name or not isinstance(app_name, str) or not app_name.strip():
                return RiskAssessment(False, "medium", "缺少应用程序名称参数", False)
            trimmed = app_name.strip()
            if "/" in trimmed or "\\" in trimmed:
                return RiskAssessment(False, "high", "应用程序名称不应包含路径分隔符", False)
            if re.search(r"\r|\n|\0", trimmed) or re.search(r"[;&|<>]", trimmed):
                return RiskAssessment(False, "high", "应用程序名称包含不安全字符", False)
            return RiskAssessment(True, "low", "ok", False)

        if name == "write_file":
            file_path = args.get("path")
            content = args.get("content")
            if not file_path or not isinstance(file_path, str):
                return RiskAssessment(False, "high", "缺少文件路径参数", True)
            if content is None or not isinstance(content, str):
                return RiskAssessment(False, "high", "缺少文件内容参数", True)
            path_check = self._is_path_allowed(file_path)
            if not path_check.allowed:
                return path_check
            return RiskAssessment(True, "medium", "ok", True)

        if name == "file_control":
            operation = args.get("operation")
            path_arg = args.get("path")
            if operation not in {"list", "read", "move", "copy", "delete", "mkdir"}:
                return RiskAssessment(False, "high", "文件操作类型不支持或缺失", True)
            path_check = self._is_path_allowed(path_arg)
            if not path_check.allowed:
                return path_check
            dst = args.get("destination")
            if dst:
                dst_check = self._is_path_allowed(dst)
                if not dst_check.allowed:
                    return RiskAssessment(False, "high", f"目标路径不安全: {dst_check.reason}", True)
            return RiskAssessment(True, "high", "ok", True)

        if name == "write_run_code":
            language = args.get("language")
            code = args.get("code")
            run = args.get("run")
            if language not in {"javascript", "python", "bash"}:
                return RiskAssessment(False, "high", "代码语言不支持或缺失", True)
            if not code or not isinstance(code, str):
                return RiskAssessment(False, "high", "代码内容无效或缺失", True)
            if run is not None and not isinstance(run, bool):
                return RiskAssessment(False, "high", "run 参数必须是布尔值", True)
            return RiskAssessment(True, "high", "ok", True)

        if name == "send_message":
            target = args.get("target")
            content = args.get("content")
            if not target or not isinstance(target, str):
                return RiskAssessment(False, "medium", "缺少消息目标参数", True)
            if not content or not isinstance(content, str):
                return RiskAssessment(False, "medium", "缺少消息内容参数", True)
            return RiskAssessment(True, "medium", "ok", True)

        if name == "music_ui":
            player = args.get("player")
            action = args.get("action")
            if player not in {"kugou", "apple_music"}:
                return RiskAssessment(False, "high", "player 仅支持 kugou / apple_music", True)
            if action not in {"favorites_first", "playlist", "search"}:
                return RiskAssessment(False, "high", "action 仅支持 favorites_first / playlist / search", True)

            # playlist/search 需要 query
            query = args.get("query")
            if action in {"playlist", "search"} and (not isinstance(query, str) or not query.strip()):
                return RiskAssessment(False, "high", "playlist/search 需要提供 query", True)

            # UI 自动化属于高风险（会真实点击屏幕）
            return RiskAssessment(True, "high", "ok", True)

        if name == "media_control":
            action = args.get("action")
            if action not in {
                "play_pause",
                "next",
                "previous",
                "volume_up",
                "volume_down",
                "mute",
                "get_now_playing",
                "set_volume_delta",
            }:
                return RiskAssessment(False, "medium", "不支持的 media_control.action", False)

            if action == "set_volume_delta":
                delta = args.get("delta")
                if delta is None:
                    return RiskAssessment(False, "medium", "set_volume_delta 需要 delta 参数", False)
                try:
                    int(delta)
                except Exception:
                    return RiskAssessment(False, "medium", "delta 必须是整数", False)

            return RiskAssessment(True, "low", "ok", False)

        if name == "run_tests":
            cwd = args.get("cwd")
            command = args.get("command")
            if not cwd or not isinstance(cwd, str):
                return RiskAssessment(False, "high", "缺少工作目录参数", True)
            if not command or not isinstance(command, str):
                return RiskAssessment(False, "high", "缺少测试命令参数", True)
            cwd_check = self._is_path_allowed(cwd)
            if not cwd_check.allowed:
                return RiskAssessment(False, "high", f"工作目录不安全: {cwd_check.reason}", True)
            return RiskAssessment(True, "high", "ok", True)

        if name == "execute_workflow":
            steps = args.get("steps")
            if not isinstance(steps, list) or not steps:
                return RiskAssessment(False, "high", "steps 必须是非空数组", True)

            # 汇总每一步的风险：任一步不允许则整体不允许；任一步需要确认则整体需要确认。
            max_level = "low"
            need_confirm = False
            for step in steps[:10]:
                if not isinstance(step, dict):
                    return RiskAssessment(False, "high", "steps 元素必须是对象", True)
                step_name = step.get("name")
                step_args = step.get("arguments") if isinstance(step.get("arguments"), dict) else {}
                if not isinstance(step_name, str) or not step_name:
                    return RiskAssessment(False, "high", "steps.name 无效", True)
                if step_name == "execute_workflow":
                    return RiskAssessment(False, "high", "不允许嵌套 execute_workflow", True)

                r = self.validate_tool_call({"name": step_name, "arguments": step_args}, allow_local_control=allow_local_control)
                if not r.allowed:
                    return RiskAssessment(False, r.risk_level, f"工作流步骤被阻止: {r.reason}", r.requires_confirmation)

                # risk level merge
                order = {"low": 0, "medium": 1, "high": 2}
                if order.get(r.risk_level, 2) > order.get(max_level, 0):
                    max_level = r.risk_level
                need_confirm = need_confirm or r.requires_confirmation

            return RiskAssessment(True, max_level, "ok", need_confirm)

        # 默认风险表
        risk_table = {
            "play_music": ("low", False),
            "stop_music": ("low", False),
            "open_app": ("low", False),
            "write_article": ("medium", False),
        }
        if name in risk_table:
            level, confirm = risk_table[name]
            return RiskAssessment(True, level, "ok", confirm)

        return RiskAssessment(False, "high", f"未知的工具类型: {name}", False)

    def generate_confirmation_request(self, tool_call: Dict[str, Any], risk: RiskAssessment) -> Dict[str, Any]:
        # 保持与 Node 侧 SafetyService.generateConfirmationRequest 的 schema 兼容。
        import time
        import random

        confirmation_id = f"confirm_{int(time.time() * 1000)}_{random.randint(1000, 9999)}"
        name = tool_call.get("name")
        args = tool_call.get("arguments") or {}

        suggestions: List[str] = []
        if name == "write_file":
            suggestions.append(f"将写入文件: {args.get('path')}")
        elif name == "send_message":
            suggestions.append(f"将发送消息给: {args.get('target')}")
        elif name == "write_run_code":
            suggestions.append("此操作可能执行本地代码，请确认是否继续")
        elif name == "file_control":
            suggestions.append("此操作可能影响本地文件，请确认是否继续")
        elif name == "music_ui":
            suggestions.append("此操作将通过 UI 自动化点击屏幕并控制播放器，请确认是否继续")

        return {
            "id": confirmation_id,
            "toolCall": {
                "id": tool_call.get("id"),
                "name": name,
                "arguments": args,
            },
            "riskLevel": risk.risk_level,
            "summary": self.summarize_tool_call(name, args),
            "reason": risk.reason,
            "suggestions": suggestions,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "expiresAt": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(time.time() + 300)),
        }

    def summarize_tool_call(self, name: str, args: Dict[str, Any]) -> str:
        if name == "open_app":
            return f"打开应用程序: {args.get('name')}"
        if name == "write_file":
            return f"写文件: {args.get('path')}"
        if name == "send_message":
            return f"发送消息给: {args.get('target')}"
        if name == "music_ui":
            return f"UI 自动化音乐控制: {args.get('player')} / {args.get('action')} / {args.get('query')}"
        if name == "media_control":
            return f"系统媒体控制: {args.get('action')}"
        if name == "write_run_code":
            run = "并运行" if args.get("run") else ""
            return f"编写{args.get('language')}代码{run}"
        if name == "file_control":
            dst = args.get("destination")
            dst_text = f" → {dst}" if dst else ""
            return f"文件操作: {args.get('operation')} {args.get('path')}{dst_text}"
        if name == "run_tests":
            return f"运行单测: {args.get('command')}（目录: {args.get('cwd')}）"
        if name == "execute_workflow":
            steps = args.get("steps")
            step_names = []
            if isinstance(steps, list):
                for s in steps[:5]:
                    if isinstance(s, dict) and isinstance(s.get("name"), str):
                        step_names.append(s.get("name"))
            joined = " -> ".join(step_names) if step_names else "(unknown)"
            return f"执行多步工作流: {joined}"
        return f"执行操作: {name}"
